### DISCUSSION TITLE
GraphQL API interface design question

### DISCUSSION THREAD

**Discussion by gopircr** [graphql-java/graphql-java#2318] [General]
We are thinking of building a new GraphQL Api that supports 2 ( or possible more in the future) different methods of delivering content - 1) Return links to CDN for the client to download or 2) Return the content as part of response in the form of a list of key/value pairs.  When a API call is made, the client is unaware if it will receive CDN links or the inline content (only one method is supported for each content bucket). Based on the response (depending on what fields are populated), client would process the response accordingly. I can think of following different approaches to handle this flow:

**Approach 1:**   Client sends the following request. Server decides to either populate url or the line under files depending on the content bucket being requested.   A response header will be added to indicate the delivery method. Does using http headers considered as a good pattern in GraphQL ?  Should delivery method be a separate field ?  Client can either look at the http header or the fields that are populated to decide how to consume the content.  There are url at multiple levels depending on how the content is zipped ( 1 zip file containing the entire bucket or zip file for each file under a content bucket).

```
  content {
        name
        version
        url {
            link
               size
               protocol
        }
        files {
           url {
               link
               size
               protocol
           }
           line {
               key
               value
           }
       }
  }
```
**Approach 2:**.   Create a separate for each type of delivery and handle the different ways of including as part of the type.
This approach also involves clients having to request both methods of delivery but the server gets to choose the supported method.   Same question as above.  Http header or a field to represent method of delivery.
Something like this

```
  content {
        name
        version
        cdn_based_delivery {
           file_urls {
               link
               size
               protocol
          }
           bucket_url {
               link
               size
               protocol
           }
       }

      files {
          line {
               key
               value
           }
       }
       
  }
```

Which approach do you think is better ? Any recommendations or inputs would be appreciated. Thanks.

**Comment by deepakguruu** [comment]
I feel using an interface and extending the return types would be better suited here.
Why is there a need for response header?
GraphQL is only going to return for either of those two types right?

**Reply by gopircr** [reply]
What kind of interface extension you are referring to ?
Response header is to indicate what is the delivery mechanism used.
API would populate either URL or files or both depending on what delivery mechanism is supported for a given bucket.

**Reply by aleksandarsusnjar** [reply]
I agree with @deepakguruu. You can have one `DeliverableFile` interface with two object types implementing/extending it, `CDNDeliverableFile` and `EmbeddableFile` -- choose your names). The client still must specify both if it can handle both, but you get to have the client express its abilities this way. The server gets to choose the actual delivery type.

Say:
```GraphQL

type Content {
    name: String!
    version: String
    files: [DeliverableFile!]!
}

interface DeliverableFile {
    id: ID!
    name: String!
}

type CDNDeliverableFile implements DeliverableFile {
    id: ID!
    name: String!
    file_url {
        link: String!
        size: Int!
        protocol: String!
    }
    bucket_url {
        link: String!
        size: Int!
        protocol: String!
    }
}

type EmbeddableFile implements DeliverableFile {
    id: ID!
    name: String!
    lines {
        key: String!
        value: String!
    }
}
```

... then your client can request it such as:

```GraphQL
content {
    files {
        ... on CDNDeliverableFile {
           file_url {
               link
               size
               protocol
          }
           bucket_url {
               link
               size
               protocol
           }
        }
        ... on EmbeddableFile {
            lines {
                key
                value
            }
        }
    }
}
```
